eqcorrscan/utils/correlate.py

```diff
@@ -869,6 +873,159 @@ def fftw_multi_normxcorr(template_array, stream_array, pad_array, seed_ids,
         stream_array[i] *= multipliers[x]
     return cccs, used_chans
 
+
+# ------------------------------- FastMatchedFilter Wrapper
+
+def _run_fmf_xcorr(template_arr, data_arr, weights, pads, arch, step=1):
+    try:
+        import fast_matched_filter
+    except ImportError:
+        raise ImportError("FastMatchedFilter is not available")
+    if version.parse(fast_matched_filter.__version__) >= MIN_FMF_VERSION:
+        from fast_matched_filter import matched_filter as fmf
+    else:
+        raise ImportError(f"FMF version {fast_matched_filter.__version__} "
+                          f"must be >= {MIN_FMF_VERSION}")
+    # Demean
+    template_arr -= template_arr.mean(axis=-1, keepdims=True)
+    data_arr -= data_arr.mean(axis=-1, keepdims=True)
+
+    multipliers = []
+    for x in range(data_arr.shape[0]):
+        # Check that stream is non-zero and above variance threshold
+        if not np.all(data_arr[x] == 0) and np.var(data_arr[x]) < 1e-8:
+            # Apply gain
+            data_arr[x] *= MULTIPLIER
+            Logger.warning(f"Low variance found for {x}, applying gain "
+                           "to stabilise correlations")
+            multipliers.append(MULTIPLIER)
+        else:
+            multipliers.append(1)
+
+    cccsums = fmf(
+        templates=template_arr, weights=weights, moveouts=pads,
+        data=data_arr, step=step, arch=arch, normalize="full")
+    # Remove gain
+    for x in range(data_arr.shape[0]):
+        data_arr[x] *= multipliers[x]
+
+    return cccsums
+
+
+@register_array_xcorr("fmf")
+def fmf_xcorr(templates, stream, pads, arch="precise", *args, **kwargs):
+    """
+    Compute cross-correlations in the time-domain using the FMF routine.
+
+    :param templates: 2D Array of templates
+    :type templates: np.ndarray
+    :param stream: 1D array of continuous data
+    :type stream: np.ndarray
+    :param pads: List of ints of pad lengths in the same order as templates
+    :type pads: list
+    :param arch: GPU or CPU to run on GPU or CPU respectively
+    type arch: str
+
+    :return: np.ndarray of cross-correlations
+    :return: np.ndarray channels used
+    """
+    assert templates.ndim == 2, "Templates must be 2D"
+    assert stream.ndim == 1, "Stream must be 1D"
+
+    used_chans = ~np.isnan(templates).any(axis=1)
+
+    # We have to reshape to an extra dimension for FMF
+    ccc = _run_fmf_xcorr(
+        template_arr=templates.reshape(
+            (1, templates.shape[0], templates.shape[1])).swapaxes(0, 1),
+        data_arr=stream.reshape((1, stream.shape[0])),
+        weights=np.ones((1, templates.shape[0])),
+        pads=np.array([pads]),
+        arch=arch)
+
+    return ccc, used_chans
+
+
+@fmf_xcorr.register("stream_xcorr")
+@fmf_xcorr.register("concurrent")
+def _fmf_gpu(templates, stream, *args, **kwargs):
+    """
+    Thin wrapper of fmf_multi_xcorr setting arch to gpu.
+    """
+    return fmf_multi_xcorr(templates, stream, arch="gpu")
```

Comment:
So I was planning on not exposing the arch argument anywhere user-facing. `fmf_multi_xcorr` should probably be a private function in that case.
I was planning on mapping concurrency to arch so that only `stream_xcorr` and `concurrent` map to `gpu` and `multithread` and `multiprocess` map to `precise`. `arch="cpu"` shouldn't be supported - that doesn't give the same results because I haven't sorted the normalisation there. `arch="cpu"` and `arch="precise"` both target the cpu, but use different strategies for computing the denominator of the correlation equation.

I'm guessing that in your tests `arch` was generally ignored as a kwarg - it doesn't get handled by the wrapper funcs.